core/client/src/children.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+
+use std::collections::HashMap;
+use kvdb::{KeyValueDB, DBTransaction};
+use parity_codec::{Encode, Decode};
+use crate::error;
+use std::hash::Hash;
+
+
+/// Map of children blocks stored in memory for fast access.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ChildrenMap<K, V> where 
+	K: Eq + Hash + Clone + Encode + Decode,
+	V: Eq + Hash + Clone + Encode + Decode,
+{
+	storage: HashMap<K, Vec<V>>,
+}
+
+impl<K, V> ChildrenMap<K, V> where
+	K: Eq + Hash + Clone + Encode + Decode,
+	V: Eq + Hash + Clone + Encode + Decode,
+{
+	/// Creates an empty children map.
+	pub fn new() -> Self {
+		Self {
+			storage: HashMap::new(),
+		}
+	}
+
+	/// Returns the hashes of the children blocks of the block with `parent_hash`.
+	pub fn hashes(&self, db: &KeyValueDB, column: Option<u32>, prefix: &[u8], parent_hash: K) -> error::Result<Vec<V>> {
+		let mut buf = prefix.to_vec();
+		parent_hash.using_encoded(|s| buf.extend(s));
+
+		let raw_val_opt = match db.get(column, &buf[..]) {
+			Ok(raw_val_opt) => raw_val_opt,
+			Err(_) => return Err(error::ErrorKind::Backend("Error reading value from database".into()).into()),
+		};
+
+		let raw_val = match raw_val_opt {
+			Some(val) => val,
+			None => return Ok(Vec::new()),
+		};
+
+		let children: Vec<V> = match Decode::decode(&mut &raw_val[..]) {
+			Some(children) => children,
+			None => return Err(error::ErrorKind::Backend("Error decoding children".into()).into()),
+		};
+
+		Ok(children)
+	}
+
+	/// Returns the hashes of the children blocks of the block with `parent_hash`.
+	/// It doesn't read the database.
+	pub fn hashes_from_mem(&self, parent_hash: K) -> Vec<V> {
+		self.storage.get(&parent_hash).cloned().unwrap_or_default()
```

Comment:
You always `clear` the storage after calling `prepare_transaction` and you call `import` directly before calling `prepare_transaction`. So, this should *always* return an empty vector. Or do I miss something?

Reply:
I think there is a flow where you only call `import` followed by `hashes_from_mem`. Like writing and reading from "memory". I was following the pattern in the sibling `leaves` module. See the next comment.

Reply:
Where is this flow? Looking into your changes I don't see it. The only place where you use the hashmap is in your tests.

Reply:
Hmm, maybe we should split the implementation into in memory and into a database implementation? Because both do not seem to have anything in common, besides of the simple `import`.

Reply:
Sounds like a good idea. I'll do that :+1:

Reply:
it's probably a good idea to keep metadata for blocks all in one place instead of many scattered indices.

Reply:
@rphmeier could you explain in more detail what you mean?

Reply:
I think after https://github.com/paritytech/substrate/pull/1779 is merged, we should be able to implement this as I expected it. (Dropping the `HashMap`, as we just use a in memory kvdb)

Reply:
@marcio-diaz could you please do, what @svyatonik has written [here](https://github.com/paritytech/substrate/pull/1779#issuecomment-463972761). So, you can remove the hashmap and the specialization for `in_mem`.

Reply:
Ok, thanks. Now, I'm trying to figure it out where is the right place to insert the children in `core/client/db/src/lib.rs` . The current place (`backend.commit_operation`) doesn't seems to be right.

Reply:
@bkchr look at how we do it in parity-ethereum. There is a `BlockDetails` mapping block hashes to a bunch of metadata: `{children, total_difficulty, some_other_things...}`.

Reply:
@rphmeier I'll create a new issue and fix it there.